feat: insert expr support

letFunny opened the pull request:
This PR completes the support for the new insert expression syntax.

letFunny committed:
feat: insert expr support

This PR completes the support for the new insert expression syntax.

Aflynn50 commented on internal/expr/bindinputs.go:
```diff
@@ -59,12 +59,43 @@ func (tbe *TypeBoundExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
 				sqlStr.WriteString("@sqlair_" + strconv.Itoa(inputCount))
 				inputCount++
 			}
+		case *typedInsertExpr:
```
In previous PR's we've talked about moving this switch case to be methods on the `typedInputExpr`. I think we decided against it because it would have been a lot of boilerplate for not much gain. Though now, in my opinion, its probably worth doing, since this function has grown so much.

The annoying thing about putting these things in methods is that they change the input/output count as well as adding the params. Also that we are actually doing something completely different for the outputs and the inputs alongside generating the SQL (generating named params vs extracting the `arginfo.Output`s).

IMO this change should be done in a subsequent PR, since frankly, the code is a lot simpler to read/review in this form and finding the least worst way to split this into functions may take some iterations.

letFunny replied:
> The annoying thing about putting these things in methods is that they change the input/output count as well as adding the params. Also that we are actually doing something completely different for the outputs and the inputs alongside generating the SQL (generating named params vs extracting the arginfo.Outputs).

Another annoying thing here is that even though the code is very similar the types are not the same, so we'll have to create an interface or something to refactor it. I agree with you that the function is too big but I am not sure whether the remedy would be worse than the cure.

letFunny replied:
> IMO this change should be done in a subsequent PR, since frankly, the code is a lot simpler to read/review in this form and finding the least worst way to split this into functions may take some iterations.

I agree with you and we should explore it further to see if it is worth it. It is best in another PR because it would probably change a lot of code and the priority is getting inserts in.

Aflynn50 commented on internal/expr/bindinputs.go:
```diff
@@ -110,6 +141,13 @@ type typedInputExpr struct {
 	input typeinfo.Input
```
Maybe worth updating the comment on this to somehow imply that doesn't include the input expressions in INSERT statements.

letFunny replied:
I changed it to:
```
// typedInputExpr stores information about a Go value to use as a standalone query
// input.
```

Aflynn50 commented on internal/expr/bindtypes.go:
```diff
@@ -132,8 +133,35 @@ func (e *asteriskInsertExpr) String() string {
 
-func (e *asteriskInsertExpr) bindTypes(argInfo typeinfo.ArgInfo) (any, error) {
-	return nil, fmt.Errorf("insert input expression not implemented")
+// bindTypes generates a *typedInputExpr containing type information about the
```
```suggestion
// bindTypes generates a *typedInsertExpr containing type information about the
```

Aflynn50 commented on internal/expr/bindtypes.go:
```diff
@@ -150,8 +178,80 @@ func (e *columnsInsertExpr) String() string {
 
-func (e *columnsInsertExpr) bindTypes(argInfo typeinfo.ArgInfo) (any, error) {
-	return nil, fmt.Errorf("insert input expression not implemented")
+// bindTypes generates a *typedInputExpr containing type information about the
```
```suggestion
// bindTypes generates a *typedInsertExpr containing type information about the
```

Aflynn50 commented on internal/expr/bindtypes.go:
```diff
@@ -150,8 +178,80 @@ func (e *columnsInsertExpr) String() string {
+
+	// 1. We go over all the supplied types and collect all the columns that
+	// they provide.
+	var remainingMap *string
```
Maybe put this below the colToInput, having it in the middle of the two comments confused me.

letFunny replied:
Agree

Aflynn50 commented on internal/expr/bindtypes.go:
```diff
@@ -150,8 +178,80 @@ func (e *columnsInsertExpr) String() string {
+		}
+	}()
+
+	// 1. We go over all the supplied types and collect all the columns that
```
Could maybe change this comment to:
```
Work out all the columns available on the right hand side of the insert expression. For each valueAccessor on the right, save all the columns names can be found in the type it specifies
```

letFunny replied:
Agree, I think yours is more explicit

Aflynn50 commented on internal/expr/bindtypes.go:
```diff
@@ -150,8 +178,80 @@ func (e *columnsInsertExpr) String() string {
+
+	// 1. We go over all the supplied types and collect all the columns that
+	// they provide.
+	var remainingMap *string
```
Could maybe have a seperate comment about the remaining map.

letFunny replied:
Agree

Aflynn50 commented on internal/expr/bindtypes.go:
```diff
@@ -150,8 +178,80 @@ func (e *columnsInsertExpr) String() string {
+		}
+	}
+
+	// 2. We go over all the columns listed in SQL and match them against the
```
Could change it to:
```
	// 2. We go over all the columns listed on the left of the insert expression and match them against the
	// columns provided by the types on the right. If a map with an asterisk is present, we use it to supply the
	// spare columns. If it is not present, a spare column is an error.
```

letFunny replied:
Agreed and changed

Aflynn50 commented on package_test.go:
```diff
@@ -819,7 +819,23 @@ func (s *PackageSuite) TestRun(c *C) {
 	err = db.Query(nil, selectStmt, &jim).Get(&jimCheck)
 	c.Assert(err, IsNil)
-	c.Assert(jimCheck, Equals, jim)
+
```
It would be good to dogfood the new syntax in the `personAndAddressDB` function as well.

letFunny replied:
Good idea! Done in the last commit!

Aflynn50 reviewed (commented):
Excellent work

letFunny committed:
address review comments

letFunny committed:
doc.go

manadart commented on internal/expr/bindinputs.go:
```diff
@@ -59,12 +59,43 @@ func (tbe *TypeBoundExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
 				sqlStr.WriteString("@sqlair_" + strconv.Itoa(inputCount))
 				inputCount++
 			}
+		case *typedInsertExpr:
```
_Code Complete 6.3 Design and Implementation Issues_:
"Prefer polymorphism to extensive type checking".

Of course we don't have polymorphism in Go, but I feel like we could do something as mentioned.
Happy for it to come later.

Aflynn50 replied:
Yea, we were thinking the same and have planned to do it in a subsequent PR. This chunk of code is particularly annoying because it really does quite different things for the different structs.

I'll have a go at separating it out.

manadart commented on internal/typeinfo/arginfo_test.go:
```diff
@@ -61,10 +61,15 @@ func (s *typeInfoSuite) TestArgInfoStruct(c *C) {
 
+	allInputs, inputNames, err := argInfo.AllStructInputs("myStruct")
+	c.Assert(err, IsNil)
+	c.Assert(allInputs, HasLen, len(structFields))
```
After you have asserted a nil error, which you want to fail immediately to prevent later check panics, prefer to use `c.Check`, which does not jump out of the test immediately. That way you get more feedback for failure cases.

Aflynn50 replied:
Changed.

manadart reviewed (approved):
Nice job. Consider the check fixes, and we're good to go.

Aflynn50 committed:
Change some Asserts to Checks

Change to Check where the table drive test can continue on a incorrect
value rather than stopping
